### PR TITLE
Visualization Framework, Antsforaging/UI, Boids_UI

### PR DESCRIPTION
**Pull request by Carbonhell**
This big pull request adds the initial implementation of the Visualization Framework, an abstraction that allows a simpler use of Amethyst for the visualization of Agent Based Simulations based on the RustAB engine. To describe the features proposed within the framework, the existing boids example was implemented with a graphical UI through this framework, and the outdated antsforaging examples have been updated to both support the new engine changes and to adapt to the proposed framework. 

**Current features:**

- Generation of the required systems to step the schedule for each frame while keeping track of the required info for the visualization layer;
- Helper struct (sprite_render_factory) to allow for the lazy initialization of emoji sprites, that can be used to quickly represent agents within a simulation visualization, for example to build quick prototypes;
- Visualization state which can accept a struct that implements a special on_init method where the creation of the various agents is supposed to be placed in (unlike CLI based simulations, where usually the generation is done in the main function of the simulation: the main function of UI-enabled simulations will simply generate the Amethyst App through the façade offered by the framework proposed, so that the developer simply has to specify the essential info such as the window size, background color and the config/asset path if needed)

**Issues:**

- The renderer system currently doesn't loop Render trait objects through par_join (in a parallel way offered by Rayon), due to an incompatibility with par_join required trait bounds and the Amethyst Transform component storage, FlaggedStorage (The issue is still quite unclear, I've asked in the official Amethyst channels for futher guidance)
- In the antsforaging simulation, adding visualization for the pheromones added quite a computing load, which causes a serious slowdown which is extremely noticeable on a 500x500 grid with 500 agents. The previous issue might be a cause, but right now each pheromone has to be represented by a Specs entity, which means 250.000 entities to represents pheromones. As efficiently as I've implemented them, the issue obviously still stands, and it might require a futher expansion of the visualization layer for this specific type of problem, perhaps by checking out packages such as amethyst_tiles or going low-level with the Rendy renderer directly to avoid representing the pheromones as entities at all.
- The visualization framework currently supports the full automation of only a single type of agent, meaning the Renderer Subsystem will call update() only on the single generic used while creating the Visualization struct. This limitation is necessary because Render trait objects must implement Component, Components must be stored in Storages and Storages must know how much memory to allocate, thus full generalization might not be possible, but I may have overlooked something and futher examination is highly suggested.

(The PR changes 934 files, but most of them are the emoji assets added within the visualization assets folder.)

### CODE REVIEW

File: examples/antsforaging/antsforaging_ui/main.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+fn main() -> amethyst::Result<()> {
+    let app_root = application_root_dir()?
+        .join("examples")
+        .join("antsforaging")
+        .join("antsforaging_ui");
+    /* CLion debugger starts the built executable directly from the target folder, so the path must
+            be hardcoded like this to be able to debug simulations.
+    let app_root = PathBuf::from(
+        "C:\\Users\\frafo\\Desktop\\Tirocinio\\abm\\examples\\antsforaging\\antsforaging_ui",
```
Comment (peppekristen) [inline examples/antsforaging/antsforaging_ui/main.rs:36]: Non puoi mettere un path assoluto? Il file appartiene al progetto

File: src/visualization/visualization_state.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        let path = application_root_dir()
+            .unwrap()
+            .join("src")
+            .join("visualization")
+            .join("assets")
+            .join("emojis");
+        /* CLion debugger starts the built executable directly from the target folder, so the path must
+            be hardcoded like this to be able to debug simulations.
+        let path = PathBuf::from(
+            "C:\\Users\\frafo\\Desktop\\Tirocinio\\abm\\src\\visualization\\assets\\emojis",
```
Comment (peppekristen) [inline src/visualization/visualization_state.rs:79]: Stessa cosa di prima, facendo riferimetno ad un file del progetto non possiamo pre impostare un path?

**Review by peppekristen — APPROVED** [review]
Per me è okay, tutto chiaro e ben documentato.

**Review by MatDau — APPROVED** [review]

**Review by ddevin96 — APPROVED** [review]

File: examples/antsforaging/antsforaging_ui/main.rs (hunk 1 above)
Comment (Carbonhell) [reply examples/antsforaging/antsforaging_ui/main.rs:36]: Intendi un path relativo? Si, tenendo conto che l'esecuzione del debugger di CLion consiste nell'eseguire l'exe dalla cartella target, posso navigare in su fino alla root del progetto per poi scendere nella cartella dell'esempio, ho voluto semplicemente lasciare il mio commento come fix temporaneo nel caso in futuro si abbia la necessità di usare il debugger per questi esempi. Durante l'esecuzione "standard" di un esempio, viene usato `cargo run` che setta automaticamente la variabile d'ambiente `CARGO_MANIFEST_DIR` alla root del progetto, quindi in quel caso questo problema non è presente. Se l'eseguibile viene lanciato manualmente, l'app root di default é settata uguale alla cwd, infatti nel caso di CLion la cwd è la cartella target. Stavo pensando di fare un readme dedicato al framework di visualizzazione dove vengono esposte le funzionalità, l'architettura e gli accorgimenti da seguire per l'esecuzione corretta di un modello, come appunto usare preferibilmente cargo run, o nel caso di modelli aventi RustAB come dipendenza cargo, lanciare l'eseguibile con cwd contenente le cartelle assets e config (usate da Amethyst)

Comment (peppekristen) [reply examples/antsforaging/antsforaging_ui/main.rs:36]: Si intendevo il path relativo. In ogni caso sicuramente fare un readme apposito è sicuramente meglio, ti dovrebbe venire anche semplice visti tutti i commenti che hai all interno del codice che sono molto esplicativi.
Ad ogni modo meglio eliminare almeno la parte relativa al percorso tuo personale del pc, anche per avere un codice più corretto e leggibile

**Review by spagnuolocarmine — APPROVED** [review]

**Pull request merged**